subsys/net/lib/nrf_cloud/coap/src/coap_codec.c

```diff
@@ -0,0 +1,631 @@
+		result->unc = res._ground_fix_resp_uncertainty_int;
+	} else {
+		result->unc = (uint32_t)lround(res._ground_fix_resp_uncertainty_float);
+	}
+	return err;
+}
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+static int format_agps_custom_types_array(struct nrf_modem_gnss_agps_data_frame const *const req,
+	int *array, int array_len)
+{
+	__ASSERT_NO_MSG(req != NULL);
+	__ASSERT_NO_MSG(array != NULL);
+	__ASSERT_NO_MSG(array_len >= (NRF_CLOUD_AGPS__LAST + 1));
+
+	int pos = 0;
+
+	if (req->data_flags & NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST) {
+		array[pos++] = NRF_CLOUD_AGPS_UTC_PARAMETERS;
+	}
+	if (req->sv_mask_ephe) {
+		array[pos++] = NRF_CLOUD_AGPS_EPHEMERIDES;
+	}
+	if (req->sv_mask_alm) {
+		array[pos++] = NRF_CLOUD_AGPS_ALMANAC;
+	}
+	if (req->data_flags & NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST) {
+		array[pos++] = NRF_CLOUD_AGPS_KLOBUCHAR_CORRECTION;
+	}
+	if (req->data_flags & NRF_MODEM_GNSS_AGPS_NEQUICK_REQUEST) {
+		array[pos++] = NRF_CLOUD_AGPS_NEQUICK_CORRECTION;
+	}
+	if (req->data_flags & NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST) {
+		array[pos++] = NRF_CLOUD_AGPS_GPS_TOWS;
+		array[pos++] = NRF_CLOUD_AGPS_GPS_SYSTEM_CLOCK;
+	}
+	if (req->data_flags & NRF_MODEM_GNSS_AGPS_POSITION_REQUEST) {
+		array[pos++] = NRF_CLOUD_AGPS_LOCATION;
+	}
+	if (req->data_flags & NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST) {
+		array[pos++] = NRF_CLOUD_AGPS_INTEGRITY;
+	}
+
+	return pos;
+}
+
+int coap_codec_agps_encode(struct nrf_cloud_rest_agps_request const *const request,
+			   uint8_t *buf, size_t *len, enum coap_content_format fmt)
+{
+	__ASSERT_NO_MSG(request != NULL);
+	__ASSERT_NO_MSG(buf != NULL);
+	__ASSERT_NO_MSG(len != NULL);
+
+	if (fmt != COAP_CONTENT_FORMAT_APP_CBOR) {
+		LOG_ERR("Invalid format for A-GPS: %d", fmt);
+		return -ENOTSUP;
+	}
+
+	int ret;
+	struct agps_req input;
+	struct agps_req_types_ *t = &input._agps_req_types;
+	size_t out_len;
+
+	memset(&input, 0, sizeof(struct agps_req));
+	input._agps_req_eci = request->net_info->current_cell.id;
+	input._agps_req_mcc = request->net_info->current_cell.mcc;
+	input._agps_req_mnc = request->net_info->current_cell.mnc;
+	input._agps_req_tac = request->net_info->current_cell.tac;
+	if (request->type == NRF_CLOUD_REST_AGPS_REQ_CUSTOM) {
+		t->_agps_req_types_int_count = format_agps_custom_types_array(
+						request->agps_req,
+						t->_agps_req_types_int,
+						ARRAY_SIZE(t->_agps_req_types_int));
+		input._agps_req_types_present = true;
+#if defined(AGPS_DEBUG)
+		LOG_DBG("num elements: %d", t->_agps_req_types_int_count);
+		for (int i = 0; i < t->_agps_req_types_int_count; i++) {
+			LOG_DBG("  %d: %d", i, t->_agps_req_types_int[i]);
+		}
+#endif
+	} else {
+		input._agps_req_requestType._agps_req_requestType._type_choice =
+			_type__rtAssistance;
+		input._agps_req_requestType_present = true;
+	}
+	if (request->filtered) {
+		input._agps_req_filtered_present = true;
+		input._agps_req_filtered._agps_req_filtered = true;
+		if (request->mask_angle != 5) {
```

Comment:
why `5`?

Reply:
That is the default angle in the modem. There is no macro for it in the various modem headers. I'll make one locally.

Reply:
or a comment would be fine, if it is only used once.